apps/state/lib/state/server/query.ex

```diff
@@ -0,0 +1,182 @@
+defmodule State.Server.Query do
+  @moduledoc """
+  Queries a Server given a map of values.
+
+  Given a Server module and a map of key -> [value], return the values from
+  the Server where each of the keys has one of the values from the given
+  list. A key without a provided list can have any value.
+
+  An example query for schedules would look like:
+
+  %{
+     trip_id: ["1", "2"],
+     stop_id: ["3", "4"]
+  }
+
+  And the results would be the schedules on trip 1 or 2, stopping at stop 3 or 4.
+  """
+
+  @type q :: map
+  @type recordable :: struct
+  @type index :: atom
+
+  alias State.Server
+
+  @spec query(module, q | [q]) :: [recordable] when q: map, recordable: struct
+  def query(module, %{} = q) when is_atom(module) do
+    do_query(module, [q])
+  end
+
+  def query(module, [%{} | _] = qs) do
+    do_query(module, qs)
+  end
+
+  def query(_module, []) do
+    []
+  end
+
+  defmodule Result do
+    @moduledoc false
+    @enforce_keys [:module]
+    defstruct [:module, matches: [], selectors: []]
+  end
+
+  alias __MODULE__.Result
+
+  defp do_query(module, qs) do
+    qs
+    |> Enum.reduce(%Result{module: module}, &accumulate/2)
+    |> finalize()
+  end
+
+  def accumulate(q, %Result{module: module} = result) when map_size(q) > 0 do
```

Comment:
Is this function returning the union of queries when it should be returning the intersection?
e.g. if `q = %{a: [1, 2], b: [3, 4]}` we want to find matches where `(a in [1, 2]) AND (b in [3, 4])`. It seems like we're always appending each individual result to the list of matches or selectors.

Reply:
It's the union of intersections, so this function handles one of the intersections, and adds it to the growing union. For example, in State.Stop:
```elixir
  def location_type_0_ids_by_parent_ids(ids) do
    [
      %{
        id: ids,
        location_type: [0]
      },
      %{
        parent_station: ids,
        location_type: [0]
      }
    ]
    |> State.Stop.Cache.query()
    |> Enum.map(& &1.id)
  end
```

This returns `(id IN ids AND location_type = 0) OR (parent_station IN ids AND location_type = 0)`.